compiler/rustc_hir_analysis/src/check/_match.rs

```diff
@@ -259,7 +259,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 err.help("consider adding an `else` block that evaluates to the expected type");
                 error = true;
             },
-            ret_reason.is_none(),
```

Comment:
For reference, this was originally introduced in ffa40cb45c195e317602437d8a40178e72807d46.

Reply:
I've noticed that commit. 
But the actual parameter of my fix will affect `label_expression_as_expected` in function `fn coerce_inner` which will reverse the expected type and found type when `label_expression_as_expected`  being true.
Before my fix, `label_expression_as_expected` being `false `when the parent of the `if `statement is function, which was wrong. Because we can't reverse the expected type and found type no matter what the parent of the `if` statement is.
I can't think of any cases of `if` statement without an `else` branch that needs to reverse the expected type and found type. If you can think of one, please let me know, I would investigate more.
The expected result of following test cases were wrong as you can tell from my fix.
```
src/test/ui/expr/if/if-without-else-result.rs
src/test/ui/expr/if/issue-4201.rs
src/test/ui/issues/issue-19991.rs
src/test/ui/async-await/issue-66387-if-without-else.rs
```
For example `src/test/ui/expr/if/issue-4201.rs`
apparently, the compiler should emit 
`expected integer, found `()``  but we emit `expected `()`, found integer`, which was reversed by mistake.